Request help creating coverage after running test executable from homebrew

I am building a test binary (gtest) as part of a homebrew build. I would like to generate a coverage report from this and pass it on to codecov.

I have this working in more simple situations, and even a similar one that was in macports; however I am removing macports and need to get coverage for macos on homebrew now. https://github.com/LizardByte/Sunshine/pull/3506

This is the step in my workflow:

```yml
      - name: Generate gcov report
        # any except canceled or skipped
        if: >-
          always() &&
          matrix.release != true &&
          (steps.test.outcome == 'success' || steps.test.outcome == 'failure')
        id: test_report
        run: |
          # if linux
          if [ "${{ runner.os }}" == "Linux" ]; then
            prefix="/tmp"
          else
            prefix="/private/tmp"
          fi

          brew_dir=$(find ${prefix} -type d -name "sunshine-*" -maxdepth 1 2>/dev/null)
          cp -rf $brew_dir/build/ ./build/
          cd build
          ls -Ra

          ${{ steps.python.outputs.python-path }} -m pip install gcovr
          ${{ steps.python.outputs.python-path }} -m gcovr . -r ../src \
            --exclude-noncode-lines \
            --exclude-throw-branches \
            --exclude-unreachable-branches \
            --verbose \
            --xml-pretty \
            -o coverage.xml
```

Which results in:
https://github.com/LizardByte/Sunshine/actions/runs/12591633444/job/35095060731?pr=3506
```txt
.
..
CMakeCache.txt
CMakeFiles
Makefile
cmake_install.cmake
sunshine.rb
tests

./CMakeFiles:
.
..
3.31.2
CMakeConfigureLog.yaml
CMakeDirectoryInformation.cmake
CMakeScratch
Makefile.cmake
Makefile2
TargetDirectories.txt
cmake.check_cache
pkgRedirects
progress.marks

./CMakeFiles/3.31.2:
.
..
CMakeCCompiler.cmake
CMakeCXXCompiler.cmake
CMakeDetermineCompilerABI_C.bin
CMakeDetermineCompilerABI_CXX.bin
CMakeSystem.cmake
CompilerIdC
CompilerIdCXX

./CMakeFiles/3.31.2/CompilerIdC:
.
..
CMakeCCompilerId.c
CMakeCCompilerId.o
tmp

./CMakeFiles/3.31.2/CompilerIdC/tmp:
.
..

./CMakeFiles/3.31.2/CompilerIdCXX:
.
..
CMakeCXXCompilerId.cpp
CMakeCXXCompilerId.o
tmp

./CMakeFiles/3.31.2/CompilerIdCXX/tmp:
.
..

./CMakeFiles/CMakeScratch:
.
..

./CMakeFiles/pkgRedirects:
.
..

./tests:
.
..
CMakeFiles

./tests/CMakeFiles:
.
..
test_sunshine.dir

./tests/CMakeFiles/test_sunshine.dir:
.
..
__
tests_main.cpp.gcda
unit

./tests/CMakeFiles/test_sunshine.dir/__:
.
..
src
third-party

./tests/CMakeFiles/test_sunshine.dir/__/src:
.
..
audio.cpp.gcda
cbs.cpp.gcda
config.cpp.gcda
confighttp.cpp.gcda
crypto.cpp.gcda
display_device.cpp.gcda
entry_handler.cpp.gcda
file_handler.cpp.gcda
globals.cpp.gcda
httpcommon.cpp.gcda
input.cpp.gcda
logging.cpp.gcda
network.cpp.gcda
nvenc
nvhttp.cpp.gcda
platform
process.cpp.gcda
rswrapper.c.gcda
rtsp.cpp.gcda
stat_trackers.cpp.gcda
stream.cpp.gcda
system_tray.cpp.gcda
upnp.cpp.gcda
video.cpp.gcda
video_colorspace.cpp.gcda

./tests/CMakeFiles/test_sunshine.dir/__/src/nvenc:
.
..
nvenc_base.cpp.gcda
nvenc_d3d11.cpp.gcda
nvenc_d3d11_native.cpp.gcda
nvenc_d3d11_on_cuda.cpp.gcda
nvenc_utils.cpp.gcda

./tests/CMakeFiles/test_sunshine.dir/__/src/platform:
.
..
macos

./tests/CMakeFiles/test_sunshine.dir/__/src/platform/macos:
.
..
av_audio.m.gcda
av_video.m.gcda
display.mm.gcda
input.cpp.gcda
microphone.mm.gcda
misc.mm.gcda
nv12_zero_device.cpp.gcda
publish.cpp.gcda

./tests/CMakeFiles/test_sunshine.dir/__/third-party:
.
..
TPCircularBuffer
moonlight-common-c

./tests/CMakeFiles/test_sunshine.dir/__/third-party/TPCircularBuffer:
.
..
TPCircularBuffer.c.gcda

./tests/CMakeFiles/test_sunshine.dir/__/third-party/moonlight-common-c:
.
..
src

./tests/CMakeFiles/test_sunshine.dir/__/third-party/moonlight-common-c/src:
.
..
RtspParser.c.gcda

./tests/CMakeFiles/test_sunshine.dir/unit:
.
..
platform
test_audio.cpp.gcda
test_entry_handler.cpp.gcda
test_file_handler.cpp.gcda
test_httpcommon.cpp.gcda
test_logging.cpp.gcda
test_mouse.cpp.gcda
test_network.cpp.gcda
test_rswrapper.cpp.gcda
test_stream.cpp.gcda
test_video.cpp.gcda

./tests/CMakeFiles/test_sunshine.dir/unit/platform:
.
..
test_common.cpp.gcda
Collecting gcovr
  Downloading gcovr-8.2-py2.py3-none-any.whl.metadata (10 kB)
Collecting jinja2 (from gcovr)
  Downloading jinja2-3.1.5-py3-none-any.whl.metadata (2.6 kB)
Collecting lxml (from gcovr)
  Downloading lxml-5.3.0-cp311-cp311-macosx_10_9_universal2.whl.metadata (3.8 kB)
Collecting colorlog (from gcovr)
  Downloading colorlog-6.9.0-py3-none-any.whl.metadata (10 kB)
Collecting pygments>=2.13.0 (from gcovr)
  Downloading pygments-2.18.0-py3-none-any.whl.metadata (2.5 kB)
Collecting MarkupSafe>=2.0 (from jinja2->gcovr)
  Downloading MarkupSafe-3.0.2-cp311-cp311-macosx_11_0_arm64.whl.metadata (4.0 kB)
Downloading gcovr-8.2-py2.py3-none-any.whl (213 kB)
Downloading pygments-2.18.0-py3-none-any.whl (1.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 40.0 MB/s eta 0:00:00
Downloading colorlog-6.9.0-py3-none-any.whl (11 kB)
Downloading jinja2-3.1.5-py3-none-any.whl (134 kB)
Downloading lxml-5.3.0-cp311-cp311-macosx_10_9_universal2.whl (8.1 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.1/8.1 MB 109.9 MB/s eta 0:00:00
Downloading MarkupSafe-3.0.2-cp311-cp311-macosx_11_0_arm64.whl (12 kB)
Installing collected packages: pygments, MarkupSafe, lxml, colorlog, jinja2, gcovr
Successfully installed MarkupSafe-3.0.2 colorlog-6.9.0 gcovr-8.2 jinja2-3.1.5 lxml-5.3.0 pygments-2.18.0
(DEBUG) File system is case insensitive.

(DEBUG) Filters for --root: (1)

(DEBUG)  - re.compile('^/Users/runner/work/Sunshine/Sunshine/src/')

(DEBUG) Filters for --filter: (1)

(DEBUG)  - DirectoryPrefixFilter(/Users/runner/work/Sunshine/Sunshine/src/)

(DEBUG) Filters for --exclude: (0)

(DEBUG) Filters for --gcov-filter: (1)

(DEBUG)  - AlwaysMatchFilter()

(DEBUG) Filters for --gcov-exclude: (0)

(DEBUG) Filters for --gcov-exclude-directories: (0)

(DEBUG) Filters for --exclude-function: (0)

(INFO) Reading coverage data...

(DEBUG) Scanning directory . for gcda/gcno files...

(DEBUG) Found 50 files (and will process 50)

(DEBUG) Pool started with 1 threads

(DEBUG) Processing file: /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda

(DEBUG) Running gcov: 'gcov --help' in '.'

(DEBUG) Running gcov: 'gcov --help-hidden' in '.'

(DEBUG) GCOV capabilities: Demangled names available.

(DEBUG) GCOV capabilities: Hashing of filenames available.

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory ../build/tests/CMakeFiles/test_sunshine.dir/unit' in '/Users/runner/work/Sunshine/Sunshine/src'

(DEBUG) GCOV return code was 0, stderr was:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory .' in '/Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit'

(DEBUG) GCOV return code was 0, stderr was:
./test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory unit' in '/Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir'

(DEBUG) GCOV return code was 0, stderr was:
unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory test_sunshine.dir/unit' in '/Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles'

(DEBUG) GCOV return code was 0, stderr was:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory CMakeFiles/test_sunshine.dir/unit' in '/Users/runner/work/Sunshine/Sunshine/build/tests'

(DEBUG) GCOV return code was 0, stderr was:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory tests/CMakeFiles/test_sunshine.dir/unit' in '/Users/runner/work/Sunshine/Sunshine/build'

(DEBUG) GCOV return code was 0, stderr was:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory build/tests/CMakeFiles/test_sunshine.dir/unit' in '/Users/runner/work/Sunshine/Sunshine'

(DEBUG) GCOV return code was 0, stderr was:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit' in '/Users/runner/work/Sunshine'

(DEBUG) GCOV return code was 0, stderr was:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit' in '/Users/runner/work'

(DEBUG) GCOV return code was 0, stderr was:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit' in '/Users/runner'

(DEBUG) GCOV return code was 0, stderr was:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit' in '/Users'

(DEBUG) GCOV return code was 0, stderr was:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(DEBUG) Running gcov: 'gcov /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda --branch-counts --branch-probabilities --all-blocks --demangled-names --hash-filenames --object-directory Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit' in '/'

(DEBUG) GCOV return code was 0, stderr was:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory
<<

(ERROR) GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.
Error: (ERROR) GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 81, in worker
    work(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 597, in process_datafile
    raise RuntimeError(errors_output)
RuntimeError: GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 81, in worker
    work(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 597, in process_datafile
    raise RuntimeError(errors_output)
RuntimeError: GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.
(ERROR) Error occurred while reading reports:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/__main__.py", line 358, in main
    covdata: CovData = gcovr_formats.read_reports(options)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/__init__.py", line 85, in read_reports
    covdata = GcovHandler(options).read_report()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/__init__.py", line 215, in read_report
    return read_report(self.options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 96, in read_report
    with Workers(
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 177, in __exit__
    self.wait()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 103, in read_report
    contexts = pool.wait()
               ^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 81, in worker
    work(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 597, in process_datafile
    raise RuntimeError(errors_output)
RuntimeError: GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.

GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.
Error: (ERROR) Error occurred while reading reports:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/__main__.py", line 358, in main
    covdata: CovData = gcovr_formats.read_reports(options)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/__init__.py", line 85, in read_reports
    covdata = GcovHandler(options).read_report()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/__init__.py", line 215, in read_report
    return read_report(self.options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 96, in read_report
    with Workers(
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 177, in __exit__
    self.wait()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 103, in read_report
    contexts = pool.wait()
               ^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 168, in wait
    raise self.exceptions[0][1]
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/workers.py", line 81, in worker
    work(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/gcovr/formats/gcov/read.py", line 597, in process_datafile
    raise RuntimeError(errors_output)
RuntimeError: GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.

GCOV produced the following errors processing /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcda:
	In directory /Users/runner/work/Sunshine/Sunshine/src:
../build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit:
./test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir:
unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles:
test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build/tests:
CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine/build:
tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine/Sunshine:
build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work/Sunshine:
Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner/work:
Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users/runner:
work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /Users:
runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	In directory /:
Users/runner/work/Sunshine/Sunshine/build/tests/CMakeFiles/test_sunshine.dir/unit/test_file_handler.cpp.gcno: No such file or directory

	(gcovr could not infer a working directory that resolved it.)
To ignore this error use option --gcov-ignore-errors=no_working_dir_found.
Error: Process completed with exit code 64.
```

Here is a previous run where I did a find operation to see where the gcda/gcno files are located, although there are no gcno files. https://github.com/LizardByte/Sunshine/actions/runs/12590322120/job/35091647502#step:9:96

I think I'm missing something simple, but not sure what it is. Any guidance would be greatly appreciated.

--- a/.github/workflows/ci-homebrew.yml
+++ b/.github/workflows/ci-homebrew.yml
@@ -177,7 +177,8 @@ jobs:
       - name: Validate Homebrew Formula
         id: test
         if: matrix.release != true
-        uses: LizardByte/actions/actions/release_homebrew@200eaeb897a2b065a65cb6f16b41077432007490  # v2026.605.34721
+        # uses: LizardByte/actions/actions/release_homebrew@200eaeb897a2b065a65cb6f16b41077432007490  # v2026.605.34721
+        uses: LizardByte/actions/actions/release_homebrew@fix/release_homebrew/reduce-test-duplication
         with:
           actionlint_config: "---\n# empty config"
           formula_file: ${{ github.workspace }}/homebrew/sunshine.rb

--- a/.github/workflows/ci-homebrew.yml
+++ b/.github/workflows/ci-homebrew.yml
@@ -198,6 +198,7 @@ jobs:
           name: coverage-Homebrew-${{ matrix.os_name }}-${{ matrix.os_version }}
           path: |
             ${{ steps.test.outputs.testpath }}/coverage.xml
+            ${{ steps.test.outputs.testpath }}/coverage.lcov
             ${{ steps.test.outputs.testpath }}/tests/test_results.xml
           if-no-files-found: error
 

--- a/.github/workflows/ci-homebrew.yml
+++ b/.github/workflows/ci-homebrew.yml
@@ -177,8 +177,7 @@ jobs:
       - name: Validate Homebrew Formula
         id: test
         if: matrix.release != true
-        # uses: LizardByte/actions/actions/release_homebrew@200eaeb897a2b065a65cb6f16b41077432007490  # v2026.605.34721
-        uses: LizardByte/actions/actions/release_homebrew@fix/release_homebrew/reduce-test-duplication
+        uses: LizardByte/actions/actions/release_homebrew@a46850981292c4bbc0c41715f286ad95adaaaf4a  # v2026.625.20301
         with:
           actionlint_config: "---\n# empty config"
           formula_file: ${{ github.workspace }}/homebrew/sunshine.rb

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -219,12 +219,15 @@ jobs:
             pr: true
           - name: Homebrew-macos-14
             coverage: true
+            coverage_file: coverage.lcov
             pr: true
           - name: Homebrew-macos-15
             coverage: true
+            coverage_file: coverage.lcov
             pr: true
           - name: Homebrew-macos-26
             coverage: true
+            coverage_file: coverage.lcov
             pr: true
           - name: Homebrew-ubuntu-24.04
             coverage: true
@@ -264,7 +267,7 @@ jobs:
         with:
           disable_search: true
           fail_ci_if_error: true
-          files: ./_coverage/coverage.xml
+          files: ./_coverage/${{ matrix.coverage_file || 'coverage.xml' }}
           report_type: coverage
           flags: ${{ matrix.name }}
           token: ${{ secrets.CODECOV_TOKEN }}

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -218,13 +218,13 @@ jobs:
             coverage: true
             pr: true
           - name: Homebrew-macos-14
-            coverage: false
+            coverage: true
             pr: true
           - name: Homebrew-macos-15
-            coverage: false
+            coverage: true
             pr: true
           - name: Homebrew-macos-26
-            coverage: false
+            coverage: true
             pr: true
           - name: Homebrew-ubuntu-24.04
             coverage: true

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -343,7 +343,7 @@ jobs:
           path: homebrew
 
       - name: Upload Homebrew Beta Formula
-        uses: LizardByte/actions/actions/release_homebrew@200eaeb897a2b065a65cb6f16b41077432007490  # v2026.605.34721
+        uses: LizardByte/actions/actions/release_homebrew@a46850981292c4bbc0c41715f286ad95adaaaf4a  # v2026.625.20301
         with:
           actionlint_config: "---\n# empty config"
           formula_file: ${{ github.workspace }}/homebrew/sunshine-beta.rb

--- a/packaging/sunshine.rb
+++ b/packaging/sunshine.rb
@@ -48,7 +48,6 @@ class Sunshine < Formula
   depends_on "graphviz" => :build if build.with? "docs"
   depends_on "node" => :build
   depends_on "pkgconf" => :build
-  depends_on "gcovr" => :test
   depends_on "boost"
   depends_on "curl"
   depends_on "icu4c@78"
@@ -62,6 +61,7 @@ class Sunshine < Formula
 
   on_linux do
     depends_on GCC_FORMULA => [:build, :test]
+    depends_on "gcovr" => [:build, :test]
     depends_on "lizardbyte/homebrew/#{CUDA_FORMULA}" => :build if build.with? "cuda"
     depends_on "python3" => :build
     depends_on "at-spi2-core"
@@ -248,6 +248,62 @@ class Sunshine < Formula
     ohai "CUDA enabled with nvcc at: #{nvcc_path}"
   end
 
+  def release_homebrew_testpath
+    testpath_value = ENV.fetch("HOMEBREW_TEST_ARTIFACTS_DIR", "")
+    return Pathname.new(testpath_value) unless testpath_value.empty?
+
+    return if ENV.fetch("HOMEBREW_TEST_BOT", "") != "1"
+
+    temp_path = ENV.fetch("HOMEBREW_TEMP", "")
+    return if temp_path.empty?
+
+    Pathname.new(temp_path)/name/"test"
+  end
+
+  def ensure_artifact_exists(path)
+    odie "#{path} was not created" unless path.exist?
+  end
+
+  def run_test_suite(artifact_dir)
+    mkdir_p artifact_dir/"tests"
+    test_results = artifact_dir/"tests/test_results.xml"
+
+    system bin/"test_sunshine", "--gtest_color=yes", "--gtest_output=xml:#{test_results}"
+    ensure_artifact_exists test_results
+  end
+
+  def generate_coverage_report(artifact_dir, coverage_buildpath)
+    return unless OS.linux?
+    return if coverage_buildpath.to_s.empty?
+
+    coverage_report = artifact_dir/"coverage.xml"
+
+    cd "#{coverage_buildpath}/build" do
+      gcc_path = Formula[GCC_FORMULA]
+      gcov_executable = "#{gcc_path.opt_bin}/gcov-#{GCC_VERSION}"
+
+      system "gcovr", ".",
+        "-r", "../src",
+        "--gcov-executable", gcov_executable,
+        "--exclude-noncode-lines",
+        "--exclude-throw-branches",
+        "--exclude-unreachable-branches",
+        "--xml-pretty",
+        "-o=#{coverage_report}"
+    end
+
+    ensure_artifact_exists coverage_report
+  end
+
+  def collect_test_artifacts
+    artifact_dir = release_homebrew_testpath
+    return unless IS_UPSTREAM_REPO
+    return unless artifact_dir
+
+    run_test_suite artifact_dir
+    generate_coverage_report artifact_dir, buildpath
+  end
+
   def build_cmake_args
     args = base_cmake_args
     add_test_args(args)
@@ -279,6 +335,7 @@ class Sunshine < Formula
     setup_build_environment
     build_and_install_project
     install_platform_specific_files
+    collect_test_artifacts
   end
 
   service do
@@ -314,33 +371,14 @@ class Sunshine < Formula
     # test that the binary runs at all
     system bin/"sunshine", "--version"
 
-    if IS_UPSTREAM_REPO && ENV.fetch("HOMEBREW_BOTTLE_BUILD", "false") != "true"
-      # run the test suite
-      system bin/"test_sunshine", "--gtest_color=yes", "--gtest_output=xml:tests/test_results.xml"
-      assert_path_exists File.join(testpath, "tests", "test_results.xml")
-
-      # create gcovr report
-      buildpath = ENV.fetch("HOMEBREW_BUILDPATH", "")
-      unless buildpath.empty?
-        # Change to the source directory for gcovr to work properly
-        cd "#{buildpath}/build" do
-          # Use GCC version to match what was used during compilation
-          if OS.linux?
-            gcc_path = Formula[GCC_FORMULA]
-            gcov_executable = "#{gcc_path.opt_bin}/gcov-#{GCC_VERSION}"
-
-            system "gcovr", ".",
-              "-r", "../src",
-              "--gcov-executable", gcov_executable,
-              "--exclude-noncode-lines",
-              "--exclude-throw-branches",
-              "--exclude-unreachable-branches",
-              "--xml-pretty",
-              "-o=#{testpath}/coverage.xml"
-
-            assert_path_exists File.join(testpath, "coverage.xml")
-          end
-        end
+    if IS_UPSTREAM_REPO
+      artifact_dir = release_homebrew_testpath
+      if artifact_dir
+        assert_path_exists artifact_dir/"tests/test_results.xml"
+        assert_path_exists artifact_dir/"coverage.xml" if OS.linux?
+      elsif ENV.fetch("HOMEBREW_BOTTLE_BUILD", "false") != "true"
+        run_test_suite testpath
+        generate_coverage_report testpath, ENV.fetch("HOMEBREW_BUILDPATH", "")
       end
     end
   end

--- a/packaging/sunshine.rb
+++ b/packaging/sunshine.rb
@@ -7,6 +7,7 @@ class Sunshine < Formula
   CUDA_FORMULA = "cuda@#{CUDA_VERSION}".freeze
   GCC_VERSION = "14".freeze
   GCC_FORMULA = "gcc@#{GCC_VERSION}".freeze
+  GCOV_IGNORE_ERRORS_FLAG = "--gcov-ignore-errors".freeze
   IS_UPSTREAM_REPO = ENV.fetch("GITHUB_REPOSITORY", "") == "LizardByte/Sunshine"
 
   desc "@PROJECT_DESCRIPTION@"
@@ -45,6 +46,7 @@ class Sunshine < Formula
 
   depends_on "cmake" => :build
   depends_on "doxygen" => :build if build.with? "docs"
+  depends_on "gcovr" => [:build, :test]
   depends_on "graphviz" => :build if build.with? "docs"
   depends_on "node" => :build
   depends_on "pkgconf" => :build
@@ -61,7 +63,6 @@ class Sunshine < Formula
 
   on_linux do
     depends_on GCC_FORMULA => [:build, :test]
-    depends_on "gcovr" => [:build, :test]
     depends_on "lizardbyte/homebrew/#{CUDA_FORMULA}" => :build if build.with? "cuda"
     depends_on "python3" => :build
     depends_on "at-spi2-core"
@@ -272,19 +273,38 @@ class Sunshine < Formula
     ensure_artifact_exists test_results
   end
 
+  def coverage_gcov_executable
+    if OS.mac?
+      llvm_path = Formula["llvm"]
+      "#{llvm_path.opt_bin}/llvm-cov gcov"
+    else
+      gcc_path = Formula[GCC_FORMULA]
+      "#{gcc_path.opt_bin}/gcov-#{GCC_VERSION}"
+    end
+  end
+
+  def coverage_gcov_options
+    options = ["--gcov-executable", coverage_gcov_executable]
+    if OS.mac?
+      options += [
+        GCOV_IGNORE_ERRORS_FLAG, "source_not_found",
+        GCOV_IGNORE_ERRORS_FLAG, "output_error",
+        GCOV_IGNORE_ERRORS_FLAG, "no_working_dir_found"
+      ]
+    end
+
+    options
+  end
+
   def generate_coverage_report(artifact_dir, coverage_buildpath)
-    return unless OS.linux?
     return if coverage_buildpath.to_s.empty?
 
     coverage_report = artifact_dir/"coverage.xml"
 
     cd "#{coverage_buildpath}/build" do
-      gcc_path = Formula[GCC_FORMULA]
-      gcov_executable = "#{gcc_path.opt_bin}/gcov-#{GCC_VERSION}"
-
       system "gcovr", ".",
         "-r", "../src",
-        "--gcov-executable", gcov_executable,
+        *coverage_gcov_options,
         "--exclude-noncode-lines",
         "--exclude-throw-branches",
         "--exclude-unreachable-branches",
@@ -375,7 +395,7 @@ class Sunshine < Formula
       artifact_dir = release_homebrew_testpath
       if artifact_dir
         assert_path_exists artifact_dir/"tests/test_results.xml"
-        assert_path_exists artifact_dir/"coverage.xml" if OS.linux?
+        assert_path_exists artifact_dir/"coverage.xml"
       elsif ENV.fetch("HOMEBREW_BOTTLE_BUILD", "false") != "true"
         run_test_suite testpath
         generate_coverage_report testpath, ENV.fetch("HOMEBREW_BUILDPATH", "")

--- a/packaging/sunshine.rb
+++ b/packaging/sunshine.rb
@@ -5,6 +5,9 @@ class Sunshine < Formula
 
   CUDA_VERSION = "13.1".freeze
   CUDA_FORMULA = "cuda@#{CUDA_VERSION}".freeze
+  COVERAGE_LCOV = "coverage.lcov".freeze
+  COVERAGE_PROFDATA = "coverage.profdata".freeze
+  COVERAGE_XML = "coverage.xml".freeze
   GCC_VERSION = "14".freeze
   GCC_FORMULA = "gcc@#{GCC_VERSION}".freeze
   LLVM_PROFILE_FILE_ENV = "LLVM_PROFILE_FILE".freeze
@@ -47,7 +50,6 @@ class Sunshine < Formula
 
   depends_on "cmake" => :build
   depends_on "doxygen" => :build if build.with? "docs"
-  depends_on "gcovr" => [:build, :test]
   depends_on "graphviz" => :build if build.with? "docs"
   depends_on "node" => :build
   depends_on "pkgconf" => :build
@@ -60,6 +62,7 @@ class Sunshine < Formula
 
   on_linux do
     depends_on GCC_FORMULA => [:build, :test]
+    depends_on "gcovr" => [:build, :test]
     depends_on "lizardbyte/homebrew/#{CUDA_FORMULA}" => :build if build.with? "cuda"
     depends_on "python3" => :build
     depends_on "at-spi2-core"
@@ -279,7 +282,7 @@ class Sunshine < Formula
   end
 
   def with_llvm_profile_file(artifact_dir)
-    original_profile_file = ENV[LLVM_PROFILE_FILE_ENV]
+    original_profile_file = ENV.fetch(LLVM_PROFILE_FILE_ENV, nil)
     ENV[LLVM_PROFILE_FILE_ENV] = "#{artifact_dir}/sunshine-%p.profraw"
     yield
   ensure
@@ -299,11 +302,12 @@ class Sunshine < Formula
     Utils.safe_popen_read("xcrun", "--find", "llvm-profdata").strip
   end
 
-  def coverage_llvm_options
-    [
-      "--llvm-profdata-executable", llvm_profdata_executable,
-      "--llvm-cov-binary", bin/TEST_BINARY
-    ]
+  def llvm_cov_executable
+    Utils.safe_popen_read("xcrun", "--find", "llvm-cov").strip
+  end
+
+  def coverage_report_path(artifact_dir)
+    artifact_dir/(OS.mac? ? COVERAGE_LCOV : COVERAGE_XML)
   end
 
   def coverage_common_options(coverage_report)
@@ -319,33 +323,80 @@ class Sunshine < Formula
   def generate_coverage_report(artifact_dir, coverage_buildpath)
     return if coverage_buildpath.to_s.empty?
 
-    coverage_report = artifact_dir/"coverage.xml"
+    coverage_report = coverage_report_path(artifact_dir)
 
     if OS.mac?
-      odie "No LLVM profile data was created" if Dir["#{artifact_dir}/*.profraw"].empty?
-
-      cd coverage_buildpath.to_s do
-        system "gcovr", artifact_dir.to_s,
-          "-r", "src",
-          *coverage_llvm_options,
-          *coverage_common_options(coverage_report)
-      end
+      generate_llvm_coverage_report artifact_dir, coverage_buildpath, coverage_report
     else
-      cd "#{coverage_buildpath}/build" do
-        system "gcovr", ".",
-          "-r", "../src",
-          *coverage_gcov_options,
-          *coverage_common_options(coverage_report)
-      end
+      generate_gcov_coverage_report coverage_report, coverage_buildpath
     end
 
     ensure_artifact_exists coverage_report
-    ensure_coverage_report_has_lines coverage_report
   end
 
-  def ensure_coverage_report_has_lines(path)
+  def generate_llvm_coverage_report(artifact_dir, coverage_buildpath, coverage_report)
+    profile_files = Dir["#{artifact_dir}/*.profraw"]
+    odie "No LLVM profile data was created" if profile_files.empty?
+
+    profile_data = artifact_dir/COVERAGE_PROFDATA
+    system llvm_profdata_executable, "merge", "--sparse", "-o", profile_data.to_s, *profile_files
+    lcov = Utils.safe_popen_read(
+      llvm_cov_executable,
+      "export",
+      "-format=lcov",
+      "-instr-profile=#{profile_data}",
+      (bin/TEST_BINARY).to_s,
+    )
+    coverage_report.write lcov_for_source_files(lcov, coverage_buildpath)
+    ensure_lcov_report_has_lines coverage_report
+  end
+
+  def generate_gcov_coverage_report(coverage_report, coverage_buildpath)
+    cd "#{coverage_buildpath}/build" do
+      system "gcovr", ".",
+        "-r", "../src",
+        *coverage_gcov_options,
+        *coverage_common_options(coverage_report)
+    end
+
+    ensure_cobertura_report_has_lines coverage_report
+  end
+
+  def coverage_source_prefixes(coverage_buildpath)
+    paths = [
+      coverage_buildpath.to_s,
+      Pathname.new(coverage_buildpath.to_s).realpath.to_s,
+    ]
+    paths.uniq.map { |path| "#{path}/src/" }
+  end
+
+  def relative_lcov_record(record, source_prefixes)
+    lines = record.lines
+    source_index = lines.index { |line| line.start_with?("SF:") }
+    return unless source_index
+
+    source_path = lines[source_index].delete_prefix("SF:").strip
+    source_prefix = source_prefixes.find { |prefix| source_path.start_with?(prefix) }
+    return unless source_prefix
+
+    lines[source_index] = "SF:src/#{source_path.delete_prefix(source_prefix)}\n"
+    "#{lines.join}end_of_record\n"
+  end
+
+  def lcov_for_source_files(lcov, coverage_buildpath)
+    source_prefixes = coverage_source_prefixes(coverage_buildpath)
+    records = lcov.split("end_of_record\n")
+    records.filter_map { |record| relative_lcov_record(record, source_prefixes) }.join
+  end
+
+  def ensure_cobertura_report_has_lines(path)
     lines_valid = path.read[/lines-valid="(\d+)"/, 1].to_i
     odie "#{path} does not contain any source lines" if lines_valid.zero?
+  end
+
+  def ensure_lcov_report_has_lines(path)
+    has_lines = path.read.lines.any? { |line| line.start_with?("DA:") }
+    odie "#{path} does not contain any source lines" unless has_lines
   end
 
   def collect_test_artifacts
@@ -428,7 +479,7 @@ class Sunshine < Formula
       artifact_dir = release_homebrew_testpath
       if artifact_dir
         assert_path_exists artifact_dir/"tests/test_results.xml"
-        assert_path_exists artifact_dir/"coverage.xml"
+        assert_path_exists coverage_report_path(artifact_dir)
       elsif ENV.fetch("HOMEBREW_BOTTLE_BUILD", "false") != "true"
         run_test_suite testpath
         generate_coverage_report testpath, ENV.fetch("HOMEBREW_BUILDPATH", "")

--- a/packaging/sunshine.rb
+++ b/packaging/sunshine.rb
@@ -7,7 +7,8 @@ class Sunshine < Formula
   CUDA_FORMULA = "cuda@#{CUDA_VERSION}".freeze
   GCC_VERSION = "14".freeze
   GCC_FORMULA = "gcc@#{GCC_VERSION}".freeze
-  GCOV_IGNORE_ERRORS_FLAG = "--gcov-ignore-errors".freeze
+  LLVM_PROFILE_FILE_ENV = "LLVM_PROFILE_FILE".freeze
+  TEST_BINARY = "test_sunshine".freeze
   IS_UPSTREAM_REPO = ENV.fetch("GITHUB_REPOSITORY", "") == "LizardByte/Sunshine"
 
   desc "@PROJECT_DESCRIPTION@"
@@ -56,10 +57,6 @@ class Sunshine < Formula
   depends_on "miniupnpc"
   depends_on "openssl@3"
   depends_on "opus"
-
-  on_macos do
-    depends_on "llvm" => [:build, :test]
-  end
 
   on_linux do
     depends_on GCC_FORMULA => [:build, :test]
@@ -184,6 +181,7 @@ class Sunshine < Formula
   def add_test_args(args)
     if IS_UPSTREAM_REPO
       args << "-DBUILD_TESTS=ON"
+      args << "-DSUNSHINE_LLVM_COVERAGE=ON" if OS.mac?
       ohai "Building tests: enabled"
     else
       args << "-DBUILD_TESTS=OFF"
@@ -269,31 +267,53 @@ class Sunshine < Formula
     mkdir_p artifact_dir/"tests"
     test_results = artifact_dir/"tests/test_results.xml"
 
-    system bin/"test_sunshine", "--gtest_color=yes", "--gtest_output=xml:#{test_results}"
+    if OS.mac?
+      with_llvm_profile_file(artifact_dir) do
+        system bin/TEST_BINARY, "--gtest_color=yes", "--gtest_output=xml:#{test_results}"
+      end
+    else
+      system bin/TEST_BINARY, "--gtest_color=yes", "--gtest_output=xml:#{test_results}"
+    end
+
     ensure_artifact_exists test_results
   end
 
-  def coverage_gcov_executable
-    if OS.mac?
-      llvm_path = Formula["llvm"]
-      "#{llvm_path.opt_bin}/llvm-cov gcov"
+  def with_llvm_profile_file(artifact_dir)
+    original_profile_file = ENV[LLVM_PROFILE_FILE_ENV]
+    ENV[LLVM_PROFILE_FILE_ENV] = "#{artifact_dir}/sunshine-%p.profraw"
+    yield
+  ensure
+    if original_profile_file
+      ENV[LLVM_PROFILE_FILE_ENV] = original_profile_file
     else
-      gcc_path = Formula[GCC_FORMULA]
-      "#{gcc_path.opt_bin}/gcov-#{GCC_VERSION}"
+      ENV.delete(LLVM_PROFILE_FILE_ENV)
     end
   end
 
   def coverage_gcov_options
-    options = ["--gcov-executable", coverage_gcov_executable]
-    if OS.mac?
-      options += [
-        GCOV_IGNORE_ERRORS_FLAG, "source_not_found",
-        GCOV_IGNORE_ERRORS_FLAG, "output_error",
-        GCOV_IGNORE_ERRORS_FLAG, "no_working_dir_found"
-      ]
-    end
+    gcc_path = Formula[GCC_FORMULA]
+    ["--gcov-executable", "#{gcc_path.opt_bin}/gcov-#{GCC_VERSION}"]
+  end
 
-    options
+  def llvm_profdata_executable
+    Utils.safe_popen_read("xcrun", "--find", "llvm-profdata").strip
+  end
+
+  def coverage_llvm_options
+    [
+      "--llvm-profdata-executable", llvm_profdata_executable,
+      "--llvm-cov-binary", bin/TEST_BINARY
+    ]
+  end
+
+  def coverage_common_options(coverage_report)
+    [
+      "--exclude-noncode-lines",
+      "--exclude-throw-branches",
+      "--exclude-unreachable-branches",
+      "--xml-pretty",
+      "-o=#{coverage_report}",
+    ]
   end
 
   def generate_coverage_report(artifact_dir, coverage_buildpath)
@@ -301,18 +321,31 @@ class Sunshine < Formula
 
     coverage_report = artifact_dir/"coverage.xml"
 
-    cd "#{coverage_buildpath}/build" do
-      system "gcovr", ".",
-        "-r", "../src",
-        *coverage_gcov_options,
-        "--exclude-noncode-lines",
-        "--exclude-throw-branches",
-        "--exclude-unreachable-branches",
-        "--xml-pretty",
-        "-o=#{coverage_report}"
+    if OS.mac?
+      odie "No LLVM profile data was created" if Dir["#{artifact_dir}/*.profraw"].empty?
+
+      cd coverage_buildpath.to_s do
+        system "gcovr", artifact_dir.to_s,
+          "-r", "src",
+          *coverage_llvm_options,
+          *coverage_common_options(coverage_report)
+      end
+    else
+      cd "#{coverage_buildpath}/build" do
+        system "gcovr", ".",
+          "-r", "../src",
+          *coverage_gcov_options,
+          *coverage_common_options(coverage_report)
+      end
     end
 
     ensure_artifact_exists coverage_report
+    ensure_coverage_report_has_lines coverage_report
+  end
+
+  def ensure_coverage_report_has_lines(path)
+    lines_valid = path.read[/lines-valid="(\d+)"/, 1].to_i
+    odie "#{path} does not contain any source lines" if lines_valid.zero?
   end
 
   def collect_test_artifacts
@@ -343,7 +376,7 @@ class Sunshine < Formula
   end
 
   def install_platform_specific_files
-    bin.install "build/tests/test_sunshine" if IS_UPSTREAM_REPO
+    bin.install "build/tests/#{TEST_BINARY}" if IS_UPSTREAM_REPO
 
     # codesign the binary on intel macs
     system "codesign", "-s", "-", "--force", "--deep", bin/"sunshine" if OS.mac? && Hardware::CPU.intel?

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -23,8 +23,14 @@ include_directories("${GTEST_SOURCE_DIR}/googletest/include" "${GTEST_SOURCE_DIR
 
 # coverage
 # https://gcovr.com/en/stable/guide/compiling.html#compiler-options
-set(CMAKE_CXX_FLAGS "-fprofile-arcs -ftest-coverage -ggdb -O0")
-set(CMAKE_C_FLAGS "-fprofile-arcs -ftest-coverage -ggdb -O0")
+option(SUNSHINE_LLVM_COVERAGE "Use LLVM source-based coverage" OFF)
+if(SUNSHINE_LLVM_COVERAGE)
+    add_compile_options(-fprofile-instr-generate -fcoverage-mapping -ggdb -O0)
+    add_link_options(-fprofile-instr-generate)
+else()
+    set(CMAKE_CXX_FLAGS "-fprofile-arcs -ftest-coverage -ggdb -O0")
+    set(CMAKE_C_FLAGS "-fprofile-arcs -ftest-coverage -ggdb -O0")
+endif()
 
 # Find the correct libgcov library path matching the gcc compiler version
 # This ensures the test executable links against the same libgcov version used during compilation

